climada/engine/uncertainty.py

```diff
@@ -0,0 +1,592 @@
+
+        self.rp = rp
+        self.calc_eai_exp = calc_eai_exp
+        self.calc_at_event = calc_at_event
+
+        if self.params.empty:
+            self.make_sobol_sample(N, calc_second_order=calc_second_order)
+        if self.aai_freq.empty:
+            self.calc_impact_distribution(rp=rp,
+                                          calc_eai_exp=calc_eai_exp,
+                                          calc_at_event=calc_at_event
+                                          )
+
+        sobol_analysis = {}
+        for imp_out in self.aai_freq:
+            Y = self.aai_freq[imp_out].to_numpy()
+            si_sobol = sobol.analyze(self.problem, Y, **kwargs)
+            sobol_analysis.update({imp_out: si_sobol})
+            for si_list in si_sobol.values():
+                if np.any(np.array(si_list)<0):
+                    LOGGER.warning("There is at least one negative sobol " +
+                        "index. Consider using more samples or using another "+
+                        "sensitivity analysis method." +
+                        "See https://github.com/SALib/SALib/issues/109")
+                    continue
+
+        self.sensitivity = sobol_analysis
+
+        return sobol_analysis
+
+
+    def calc_cost_benefit_sobol_sensitivity(self):
+        raise NotImplementedError()
+
+
+    def est_comp_time(self):
+        """
+        Estimate the computation time
+
+        Returns
+        -------
+        None.
+
+        """
+        raise NotImplementedError()
+
+
+    def calc_impact_distribution(self,
+                                 rp=None,
+                                 calc_eai_exp=False,
+                                 calc_at_event=False,
+                                 ):
+        """
+        Computes the impact for each of the parameters set defined in
+        uncertainty.params.
+
+        By default, the aggregated average annual impact
+        (impact.aai_agg) and the excees impact at return periods (rp) is
+        computed and stored in self.aai_freq. Optionally, the impact at
+        each centroid location is computed (this may require a larger
+        amount of memory if the number of centroids is large).
+
+        Parameters
+        ----------
+        rp : list(int), optional
+            Return period in years to be computed.
+            The default is [5, 10, 20, 50, 100, 250].
+        calc_eai_exp : boolean, optional
+            Toggle computation of the impact at each centroid location.
+            The default is False.
+        calc_at_event : boolean, optional
+            Toggle computation of the impact for each event.
+            The default is False.
+
+        Returns
+        -------
+        None.
+
+        """
+
+        if rp is None:
+            rp=[5, 10, 20, 50, 100, 250]
+
+        aai_agg_list = []
+        freq_curve_list = []
+        if calc_eai_exp:
+            eai_exp_list = []
+        if calc_at_event:
+            at_event_list = []
+
+        self.rp = rp
```

Comment:
I think this is a repetation; they are aleady specified at lines 414-416, before calling calc_impact_distribution

Reply:
This is intended. My current idea is that the workflow would be:

1. define the UncVar()
2. make the sample 
3. compute the impact/cost benefit distributions
4. compute the sensitivity indices

Reply:
Yes but it seems to me that the lines
`        self.rp = rp
        self.calc_eai_exp = calc_eai_exp
        self.calc_at_event = calc_at_event
`
are both in _calc_impact_sobol_sensitivity_ and _calc_impact_distribution_, with the latter being called once and inside the former; so those assignments are called twice. It is true though that, in principle, the second function could be used as standalone depending on the application, so it may be good to leave it like that.

Reply:
Yes,  I agree, that is a good summary. I will do the refactoring and then we can decide whether this is a good idea or not.